content/blog/2024/transformer_progress.md

```diff
@@ -0,0 +1,53 @@
+
+#### Positional Encodings and Smolgen
+
+A major respect which transformers have been improved in other domains is in altering the attention weights to downweigh connections between tokens which are far away from each other. The dot-product similarity metric can struggle to detect nearby tokens. Techniques like [RoPE](https://arxiv.org/abs/2104.09864) and [relative positional encodings](https://arxiv.org/pdf/1803.02155.pdf) in NLP addressed this issue by allowing models to decay the weight between distant tokens. Similar techniques in two dimensions have been applied to vision. One can imagine the tokens in the respective domains as being embedded in a “topology” whose structure roughly translates to Euclidean 1- and 2-space. Tokens can be viewed as having both “semantic” and “positional” information, with dot-produce similarity attention addressing the semantic component and positional embeddings addressing the positional component.
+
+Our main success relied on viewing the chess board as a special topology in which the idea of distance is focused on attributes of chess rather than Euclidean distance. Squares which are a knight’s, rook’s or bishop’s move away from each other can be viewed as tied closely together even if they are far apart on the board. One advantage of having the tokens corresponding to squares on a chessboard rather than characters in a fen is that the positional relationships between them are fixed. To allow the model to encode this information, our first iteration added a trainable bias to attention logits before the softmax, for a total of 64&times;64&times;h additional parameters per layer, where h is the head count. This allowed the model to play as if 50% larger.
+
+Our second improvement came from the idea that if the semantic relationships modeled with cosine similarity are dynamic then the positional information should be too. In a closed position, squares that are far apart should have the signal between them constrained while in an open position the signal between those squares should open up. The main idea is:
+
+ 1.	Compress the current representation of the position into a small vector 
+ 2.	For each head, generate a supplemental set of attention logits to add to those generate by dot-product attention before the softmax.
+
+To accomplish 1., we apply a linear projection to form a vector of length 32 per square, then flatten this into a vector of size 64&times;32, then apply another dense layer to extract a vector of size 256. For 2. we apply a linear projection to form a vector of size 256 for each head, then for each head we use a 256&times;4096 linear projection shared across the entire model to generate the final supplemental attention logits. Note that this description omits layer normalizations and biases, and the hidden sizes vary with model sizes. This module, which we call smolgen, allows the model to play as if it were an additional 50% larger with a ~15% decrease in throughput.
```

Comment:
I'd suggest we highlight the word __smolgen__, and maybe even link it to the README at https://github.com/Ergodice/lczero-training#smolgen